### PR TITLE
add ROADMAP.md

### PR DESCRIPTION
**Pull request by michelleN**
+ Add roadmap document to help describe near ish term goals for the project

### CODE REVIEW

**Comment by radu-matei** [comment]
This looks great! Thank you for putting this together!

The only question I have is whether we want to explicitly point out work that is in progress right now already (things like factors, or component dependencies).

**Review by radu-matei — APPROVED** [review]

**Comment by michelleN** [comment]
Cc @rajatjindal

File: ROADMAP.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# Spin Roadmap
+
+Spin is an open-source framework designed to facilitate the development and deployment of event-driven, serverless applications using WebAssembly (Wasm). Spin provides a lightweight runtime environment that encompasses Wasmtime for running fast and secure applications integrating the latest standards in the WebAssembly community. We’re excited for the growing engagement in the Spin community and for the hard earned continued development and progress in the WebAssembly community that are incorporated in the Spin project.
+
+Taking these considerations into account and the feedback for the community, the Spin project is focusing on features and improvements both as a developer tool and as an extensible runtime environment. The following is a non-exhaustive list:
```
Comment (vdice) [inline ROADMAP.md:5]: Was this meant to be "and feedback from the community"?

File: ROADMAP.md [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+3. Spin application development experience improvements
+It has become increasingly clear that there is interest in scenarios where Spin is used as a developer tool to build applications that target runtimes other than the Spin runtime (example: [NGINX Unit](https://unit.nginx.org/news/2024/fermyon-spin-rust-sdk/)). Right now, there are implicit assumptions made about the target environment both by the Spin CLI and the Spin SDKs. To support these scenarios,  there will need to be work done around modularizing SDKs and building tooling that validates a Spin application can be run with a specific target environment. Foundational work for these efforts is described in the [Spin Build Target Check SIP](https://github.com/fermyon/spin/pull/2556) and is an ongoing effort.
```
Comment (vdice) [inline ROADMAP.md:14]:
```suggestion
It has become increasingly clear that there is interest in scenarios where Spin is used as a developer tool to build applications that target runtimes other than the Spin runtime (example: [NGINX Unit](https://unit.nginx.org/news/2024/fermyon-spin-rust-sdk/)). Right now, there are implicit assumptions made about the target environment both by the Spin CLI and the Spin SDKs. To support these scenarios, there will need to be work done around modularizing SDKs and building tooling that validates a Spin application can be run with a specific target environment. Foundational work for these efforts is described in the [Spin Build Target Check SIP](https://github.com/fermyon/spin/pull/2556) and is an ongoing effort.
```

File: ROADMAP.md [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+1. Composition and polyglot development experience
+Web Assembly components unlock a new opportunity as it relates to the polyglot development experience. In Spin, we seek to enable the polyglot re-use of components, allowing developers to specify how to use components written in various languages as libraries to fulfill dependencies in their Spin components as specified in the [Component Dependencies SIP](https://github.com/fermyon/spin/pull/2543). We’ll use this as the foundational work to then iterate on the polyglot development experience.
+
+2. Building a composable runtime
+There are currently some assumptions baked into the Spin runtime about the functionality provided by underlying host environments, however it’s likely a host environment may offer a subset of the capabilities assumed in the Spin runtime today or an entirely custom set of capabilities. Modularizing the Spin runtime allows it to be more readily extendable and customizable based on host environment. This requires breaking changes and the major refactor being done in the ongoing [Spin Factors](https://github.com/fermyon/spin/pull/2518) work so this work will also be the basis for a Spin 3.0 release.
```
Comment (vdice) [inline ROADMAP.md:11]:
```suggestion
There are currently some assumptions baked into the Spin runtime about the functionality provided by underlying host environments, however it’s likely a host environment may offer a subset of the capabilities assumed in the Spin runtime today or an entirely custom set of capabilities. Modularizing the Spin runtime allows it to be more readily extendable and customizable based on host environment. This requires breaking changes and a major code refactor, so this work will also be the basis for a Spin 3.0 release. See the [Spin Factors SIP](https://github.com/fermyon/spin/pull/2518) for an overview.
```

File: ROADMAP.md [hunk 4]
```diff
@@ -0,0 +1,14 @@
+# Spin Roadmap
+
+Spin is an open-source framework designed to facilitate the development and deployment of event-driven, serverless applications using WebAssembly (Wasm). Spin provides a lightweight runtime environment that encompasses Wasmtime for running fast and secure applications integrating the latest standards in the WebAssembly community. We’re excited for the growing engagement in the Spin community and for the hard earned continued development and progress in the WebAssembly community that are incorporated in the Spin project.
+
+Taking these considerations into account and the feedback for the community, the Spin project is focusing on features and improvements both as a developer tool and as an extensible runtime environment. The following is a non-exhaustive list:
+
+1. Composition and polyglot development experience
```
Comment (vdice) [inline ROADMAP.md:7]: Do we want to insert a 'Spin 3.0' or similar header here to signify that the following three areas of change are anticipated to ship in this release?

In general, do we anticipate that this roadmap will grow to include details around subsequent releases?  If so, all minor ones as well as major?

Comment (vdice) [inline ROADMAP.md:7]: Markdown nit: Can the sub-section text be separated by a newline from the topic and/or indented?  Currently the markdown renders it all in one block:

Screenshot 2024-08-12 at 10 25 57 AM

**Review by rylev — APPROVED** [review]

File: ROADMAP.md (hunk 4 above)
Comment (kate-goldenring) [inline ROADMAP.md:7]: The rendering doesn't make it clear that there is a subject to each point. Maybe add a hyphen or colon
```suggestion
1. Composition and polyglot development experience -
```

File: ROADMAP.md [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+Web Assembly components unlock a new opportunity as it relates to the polyglot development experience. In Spin, we seek to enable the polyglot re-use of components, allowing developers to specify how to use components written in various languages as libraries to fulfill dependencies in their Spin components as specified in the [Component Dependencies SIP](https://github.com/fermyon/spin/pull/2543). We’ll use this as the foundational work to then iterate on the polyglot development experience.
```
Comment (kate-goldenring) [inline ROADMAP.md:8]:
```suggestion
WebAssembly components unlock a new opportunity as it relates to the polyglot development experience. In Spin, we seek to enable the polyglot re-use of components, allowing developers to specify how to use components written in various languages as libraries to fulfill dependencies in their Spin components as specified in the [Component Dependencies SIP](https://github.com/fermyon/spin/pull/2543). We’ll use this as the foundational work to then iterate on the polyglot development experience.
```

File: ROADMAP.md
Comment (kate-goldenring) [inline ROADMAP.md:1]: I'd add a section at the end with something along the lines of: "To granularly see which issues are in progress or on the backlog, see the spin project board (link)".

**Review by kate-goldenring — APPROVED** [review]
LGTM -- a couple syntax nits. Thanks for putting this together!

File: ROADMAP.md
Comment (michelleN) [reply ROADMAP.md:1]: I'll add this as a follow up since we don't have the project board just yet but thanks for the feedback and yea I agree.

Comment (michelleN) [reply ROADMAP.md:1]: https://github.com/fermyon/spin/issues/2712

**Pull request merged**